syntax = "proto3";
package kvstore;

service KvStore {
  rpc Put(PutRequest) returns (PutResponse);
  rpc Swap(SwapRequest) returns (SwapResponse);
  rpc Get(GetRequest) returns (GetResponse);
  rpc Scan(ScanRequest) returns (ScanResponse);
  rpc Delete(DeleteRequest) returns (DeleteResponse);
}

message PutRequest {
  string key = 1;
  string value = 2;
}

message PutResponse {
  bool found = 1;
}

message SwapRequest {
  string key = 1;
  string value = 2;
}

message SwapResponse {
  bool found = 1;
  string old_value = 2;
}

message GetRequest {
  string key = 1;
}

message GetResponse {
  bool found = 1;
  string value = 2;
}

message ScanRequest {
  string key_start = 1;
  string key_end = 2;
}

message ScanResponse {
  repeated KvEntry entries = 1;
}

message KvEntry {
  string key = 1;
  string value = 2;
}

message DeleteRequest {
  string key = 1;
}

message DeleteResponse {
  bool found = 1;
}
